### DISCUSSION TITLE
Arrow Function과 기본 Function의 차이점이 무엇일까요?

### DISCUSSION THREAD

**Discussion by BaekYuri** [wanted-9th-3team/frontend-study#8] [프론트 기술 질문 공간]
Arrow Function과 기본 Function의 차이점이 무엇일까요?

**Comment by skyhanull** [comment]
화살표함수는 es6버전에서 새로 생성된 함수형입니다.
기존의 function에 비해 길이가 짧을 시 중괄호를 생략할 수 있으며
화살표함수의 특징으로는
1. prototype이 없어서 일반함수보다 가볍게 사용할 수 있다.
=> _proto_가 있으면 그 안에 여러 속성이 들어있는데 화살표함수는 프로토타입이 없어서 상대적으로 가볍습니다.
2. arguments 사용이 불가능하다
3. 작성된 위치에 따라 this가 정적으로 바인딩 된다.
=> 화살표함수의 this는 화살표함수가 호출되는 시점과는 무관하게 선언되는 시점에 결정되며 언제나 상위 스코프의 this를 가리킨다.
라는 특징이 있습니다

**Comment by michoball** [comment]
Arrow Function 은 기존 Function을 대체하는 문법이지만 몇가지 차이점이 있습니다. 

- Arrow Function은 함수 표현식이기 때문에 기존의 선언형과 달리 호이스팅에 영향을 받지 않습니다. 

- Arrow  function의 this는 자신을 감싼 정적 범위를 가르킵니다.

**Reply by michoball** [reply]
arrow function은 익명함수로서 arrow function 으로 함수를 만들면 
‎Execution Context 안에 함수로서 들어가지 않고 변수를 선언한 것처럼 들어갑니다. 
그렇기에 호이스팅에 영향을 받지 않아 함수 선언전에 함수를 호출할 수 없습니다. 

this 에서 arrow function이 자신을 감싼 정적 범위를 가르킨다는 말이 바로 arrow function가 ‎Execution Context 에서 변수처럼 선언되기 때문입니다.

‎Execution Context  call stack에서  arrow function을 감싸고 있는 스코프 안에 변수처럼 선언되지  새로운 function으로 call stack에 쌓이지 않기 때문에 this가 자신을 감싼 정적범위를 가르킨다고 할 수 있다.

**Comment by BaekYuri** [comment]
기존 Function
- 메소드 호출시, this는 메소드를 소유하고 있는 객체를 가리킨다

Arrow Function
- 자신을 둘러싸고 있는 상위 환경의 this를 그대로 계승한다.
- 파라미터가 하나라면 소괄호 생략 가능
- return 한 줄 뿐이라면 중괄호 생략 가능

**Comment by javascriptbiggosoo** [comment]
(개인 정리 복붙, 수정하겠음)

JS를 보다 쉽게 쓰고싶다는 요구에 맞춰 ES6는 많은 편의성을 제공했습니다.
arrow function도 그런 배경에서 탄생한 것으로 regular function과 몇가지 의도적인 차별점을 두었습니다.
그 중 하나는 `this` keyword에 관한 것으로, 분명 함수가 자동으로 `this` keyword를 만든다고 했는데 놀랍게도!!
*arrow functions*는 안만듭니다??(아 왜요 ㅠㅠ)
암튼 그런 이유로 objects 안에 method를 arrow function으로 작성하면

```jsx
const jinyoung = {
  firstName: 'Jinyoung',
  greet: () => {
    console.log(this); // Window {window: Window, self: Window, document: document, name: "", location: Location, …}
    console.log(`Hey ${this.firstName}`); // Hey undefined
  },
};
jinyoung.greet();
```

`this` 키워드를 이용할 때에 **undefined**가 나올 것입니다.
이런 결과가 나오는 까닭은 함수 내에서 `this` 키워드가 없으니까 부모 스코프(이 경우 global)에서 `this`를 찾는 것이지요..
그러니까 **arrow functions**는 절대로 **methods**로 쓰지 마십시오.(외우셈)

그리고 regular function에서는 `this`의 값 작성시점으로 고정되어있지않고, functions가 호출되는 시점에 비로소 할당된다고 했는데, 아마 arrow funtion의 경우엔 그렇지 않은듯 넘겨받은 this가 있다면 계속 그걸 가르킨다?
일단 애로우펑션의 this도 object인것은 new 키워드 때문인 것 같다.
[https://stackoverflow.com/questions/72960454/i-so-confused-of-this-keyword-in-js-classes](https://stackoverflow.com/questions/72960454/i-so-confused-of-this-keyword-in-js-classes)
[https://www.udemy.com/course/the-complete-javascript-course/learn/lecture/22649229#questions/17863302](https://www.udemy.com/course/the-complete-javascript-course/learn/lecture/22649229#questions/17863302)

자 암튼 `this`를 안만든다는건 알아봤는데 왜 안만든지 설계의도를 파악해야 답답함이 해소되겠지요?
프로그래밍을 하다보면 **methods**안에 또 **function**을 써먹을 일이 있을 것이고, 거기서 또 **object**의 **properties**를 갖다 쓸 일도 있을 것입니다.
그런데 분명 (**regular**)**functions**는 자동으로 자신의 `this`를 새로 만든다고 하였지요?
**method**의 `this`와, **method** 내부의 **functions**의 `this`는 한 스코프 차이가 생기면서 각자가 가르키는게 다른 녀석이 된다 이말입니다.
과거엔 **method** 안에서 **variable**을 새로 만들고 후에 `this`를 그 **variable**에 할당하는 방식을 통해 문제를 해결해왔습니다.

- (예시코드)
    
    ```jsx
    const jinyoung = {
      year: 1994,
      calcAge: function () {
        console.log(this); // { year: 1994, calcAge: [Function] }
        console.log(2021 - this.year); // 27
        // Solution 1
        const self = this;
        const isMillenial = function () {
          console.log(self); // { year: 1994, calcAge: [Function] }
          console.log(self.year >= 1981 && self.year <= 1996); // true
          // console.log(this.year >= 1981 && this.year <= 1996);
        }
        isMillenial();
      },
    };
    jinyoung.calcAge();
    ```
    

그런데 **arrow**가 `this`를 안만드니까 세상에 이렇게 좋은 일이..

```jsx
const jinyoung = {
  year: 1994,
  calcAge: function () {
    console.log(this);
    console.log(2021 - this.year);
    // Solution 2 (=ES6 방식)
    const isMillenial = () => {
      console.log(this);
      console.log(this.year >= 1981 && this.year <= 1996);
    };
    isMillenial();
  },
};
jinyoung.calcAge();
```

**arrow function**이 자체 `this`를 만들지 않고 부모 EC것을 갖다쓰는거?
조금 배울 땐 오히려 안좋다가 많이 배우고 익숙해지면 오히려 좋음
